mqtt-streaming/src/main/scala/akka/stream/alpakka/mqtt/streaming/impl/ClientState.scala

```diff
@@ -0,0 +1,526 @@
+      override val unsubscriberPacketRouter: ActorRef[LocalPacketRouter.Request[Unsubscriber.Event]],
+      override val settings: MqttSessionSettings
+  ) extends Data(consumerPacketRouter,
+                   producerPacketRouter,
+                   subscriberPacketRouter,
+                   unsubscriberPacketRouter,
+                   settings)
+
+  sealed abstract class Event
+  final case class ConnectReceivedLocally(connect: Connect,
+                                          connectData: ConnectData,
+                                          remote: ActorRef[Source[ForwardConnectCommand, NotUsed]])
+      extends Event
+  final case class ConnAckReceivedFromRemote(connAck: ConnAck, local: ActorRef[ForwardConnAck]) extends Event
+  case object ReceiveConnAckTimeout extends Event
+  case object ConnectionLost extends Event
+  final case class DisconnectReceivedLocally(remote: ActorRef[ForwardDisconnect.type]) extends Event
+  final case class SubscribeReceivedLocally(subscribe: Subscribe,
+                                            subscribeData: Subscriber.SubscribeData,
+                                            remote: ActorRef[Subscriber.ForwardSubscribe])
+      extends Event
+  final case class PublishReceivedFromRemote(publish: Publish, local: ActorRef[Consumer.ForwardPublish.type])
+      extends Event
+  final case class PublishReceivedLocally(publish: Publish,
+                                          publishData: Producer.PublishData,
+                                          remote: ActorRef[Source[Producer.ForwardPublishingCommand, NotUsed]])
+      extends Event
+  final case class ProducerFree(topicName: String) extends Event
+  case object SendPingReqTimeout extends Event
+  final case class PingRespReceivedFromRemote(local: ActorRef[ForwardPingResp.type]) extends Event
+  final case class UnsubscribeReceivedLocally(unsubscribe: Unsubscribe,
+                                              unsubscribeData: Unsubscriber.UnsubscribeData,
+                                              remote: ActorRef[Unsubscriber.ForwardUnsubscribe])
+      extends Event
+
+  sealed abstract class Command
+  sealed abstract class ForwardConnectCommand
+  case object ForwardConnect extends ForwardConnectCommand
+  case object ForwardPingReq extends ForwardConnectCommand
+  final case class ForwardConnAck(connectData: ConnectData) extends Command
+  case object ForwardDisconnect extends Command
+  case object ForwardPingResp extends Command
+
+  // State event handling
+
+  private val ConsumerNamePrefix = "consumer-"
+  private val ProducerNamePrefix = "producer-"
+  private val SubscriberNamePrefix = "subscriber-"
+  private val UnsubscriberNamePrefix = "unsubscriber-"
+
+  def disconnected(data: Uninitialized): Behavior[Event] = Behaviors.receivePartial {
+    case (context, ConnectReceivedLocally(connect, connectData, remote)) =>
+      implicit val mat: Materializer = ActorMaterializer()(context.system)
```

Comment:
Creating a materializer that is bound to the ActorSystem from inside an actor is an anti pattern (possible memory leak). I'd recommend sharing one materializer by passing in constructor parameter or putting it in an extension. For actor bound materializer there is an open PR: https://github.com/akka/akka/pull/25568

Reply:
Thanks for the pick-up. Changed.